rpc/gnmi/gnmi-depth.md

```diff
@@ -0,0 +1,271 @@
+
+## 2. Demo model
+
+To explain the concept of the depth-based filtering, consider the following model that is used in the implementation examples throughout this document:
+
+```yang
+    container basket {
+      leaf name {
+        type string;
+      }
+
+      leaf-list contents {
+        type string;
+      }
+
+      list fruits {
+        key "name";
+
+        leaf name {
+          type string;
+        }
+
+        leaf-list colors {
+          type string;
+        }
+
+        leaf size {
+          type string;
+        }
+
+        container origin {
+          leaf country {
+            type string;
+          }
+          
+          leaf city {
+            type string;
+          }
+        }
+      }
+
+      container description {
+        leaf fabric {
+          type string;
+        }
+      
+      }
+      container broken {
+        presence "This container is broken";
+        leaf reason {
+          type string;
+        }
+      }
+    }
+```
+
+It's tree representation:
+
+```
+module: app
+  +--rw basket
+     +--rw name?          string
+     +--rw contents*      string
+     +--rw fruits* [name]
+     |  +--rw name      string
+     |  +--rw colors*   string
+     |  +--rw size?     string
+     |  +--rw origin
+     |     +--rw country?   string
+     |     +--rw city?      string
+     +--rw description
+     |  +--rw fabric?   string
+     +--rw broken!
+        +--rw reason?   string
+```
+
+We populate this data schema with the following values:
+
+```
+    basket {
+        contents [
+            fruits
+            vegetables
+        ]
+        fruits apples {
+            size XL
+            colors [
+                red
+                yellow
+            ]
+            origin {
+                country NL
+                city Amsterdam
+            }
+        }
+        fruits orange {
+            size M
+        }
+        description {
+            fabric cotton
+        }
+        broken {
+            reason "too heavy"
+        }
+    }
+```
+
+## 3. Concepts
+
+The Depth extension allows clients to specify the depth of the subtree to be returned in the response. The depth is specified as the number of levels below the specified path.
+
+The extension itself has a single field that controls the depth level:
+
+```proto
+message Depth {
+  uint32 level = 1;
+}
+```
+
+### 3.1 Depth level values
+
+#### 3.1.1 Value 0
+
+Depth value of 0 means no depth limit and behaves the same as if the extension was not specified at all.
+
+#### 3.1.2 Value 1
+
+Value of 1 means only the specified path and its direct children will be returned. See Children section for more info.
+
+#### 3.1.2 Value of N+
+
+Value of N+ where N>1 means all elements of the specified path up to N level and direct children of N-th level.
+
+### 3.2 Children nodes
+
+The Depth extension operates the value of "direct children of a schema node". What we understand by direct children:
```

Comment:
Since gNMI covers more than YANG-modelled data, then I think it is desirable to have the specification apply to data that is not YANG modelled. Particularly, we could do this by specifying scalar values, and arrays of scalar values. You can tie it to the examples in YANG if desired.

It's probably equally valid to say that any node in the tree that is a leaf node is returned, and explicitly nodes that are the root of a subtree are not returned.

Reply:
agree, will add an interpretation of this